src/vos/vos_obj.c

```diff
@@ -89,6 +177,14 @@ key_punch(struct vos_object *obj, daos_epoch_t epoch, uint32_t pm_ver,
 		if (rc != 0)
 			D_GOTO(out, rc);
 
+		if (rc == 0) {
+			rc = vos_propagate_check(obj, obj->obj_toh, ts_set,
```

Comment:
A tricky case:

    u(c/o/d/a0, 10) // committed
    u(c/o/d/a1, 15) // prepared
    p(c/o/d/a0, 20)

If the u(a1, 15) commits, the p(a0, 20) doesn't need to propagate to c/o/d; if the u(a1, 15) aborts, the p(a0, 20) does need to propagate. Hence, the p(a0, 20) needs to wait for the u(a1, 15) to finalize, just like regular reads! :( Is this the case?

Reply:
Good point.  I may need to rework a bit to move these checks down into the iterator logic instead.

Reply:
Hmm, can we guarantee the u(a1, 15) is executed ealier than p(a0, 20) (at least on one target)? If not, I'm afraid that local target won't have enough knowedge to detect this situation.

Reply:
In this case, the u(a1, 15) have already been executed when the p(a0, 20) comes in. So the empty check iteration in p(a0, 20) shall see the u(a1, 15) and its "prepared" status.